### ISSUE TITLE
Setting hostname with trailing dot fails validation of ValidLinuxHostname

### ISSUE THREAD

**Issue by markusboehme** [bottlerocket-os/bottlerocket#3260]
**Image I'm using:** Any current (Bottlerocket v1.14.2) image for the AWS platform

**What I expected to happen:** When configuring a custom DHCP Option Set for my VPC containing a domain with a trailing dot, I expected Bottlerocket instances launched in that VPC to successfully boot and join a cluster (ECS, EKS).

**What actually happened:** The boot fails because the resulting hostname fails validation:

```
[  OK  ] Finished Bottlerocket userdata configuration system.
         Starting User-specified setting generators...
[    6.681623] sundog[1561]: Error deserializing HashMap to Settings: Error deserializing scalar value: Unable to deserialize into ValidLinuxHostname: Invalid hostname 'ip-172-31-41-87.example.com.': segment is less than 1 or greater than 63 chars
[FAILED] Failed to start User-specified setting generators.
```

The validation logic of `ValidLinuxHostname` treats the trailing period as introducing another empty segment.

**How to reproduce the problem:**

1. Create a custom VPC DHCP Option Set, e.g. via `aws ec2 create-dhcp-options --dhcp-configurations '{"Key": "domain-name", "Values": ["example.com."]}'` (note the trailing dot)
2. Associate the option set with a VPC, e.g. via `aws ec2 associate-dhcp-options --dhcp-options-id ${DOPT_ID} --vpc-id ${VPC_ID}`
3. Launch a new Bottlerocket node into a subnet of the VPC and observe the boot failing with the above error in the system log

**Comment by markusboehme** [comment]
**tl;dr:** I initially wasn't sure about the best course of action here. The observation that such a DHCP Option Set can be created via the API but not via the web console added to this. However, I now think Bottlerocket should trim the trailing dot from the domain name (DHCP)/host name (IMDS) in its IMDS client.

**Details:** What's happening is that sundog invokes netdog to generate the hostname, which in turn queries IMDS `/meta-data/local-metadata` when running on AWS. Sundog tries to deserialize the received hostname into a `ValidLinuxHostname`, which fails. Since the queried IMDS route reflects the DHCP options, let's look at those.

[RFC 2132 section 3.17](https://www.rfc-editor.org/rfc/rfc2132#section-3.17) defines the domain name DHCP option. It does not mention what exactly is permissible, though. Turning to [RFC 1035 section 2.3.1](https://www.rfc-editor.org/rfc/rfc1035#section-2.3.1) on domain names forbids e.g. `.` in a label (i.e. the segments between `.`), but [section 3.1](https://www.rfc-editor.org/rfc/rfc1035#section-3.1) mentions every domain name ending in a null label. A trailing dot would result in such a null label. I'd therefore interpret the trailing dot in the domain name DHCP option to be permissible, too.

Looking at a packet capture of lease negotiation, the DHCP server is sending the domain name DHCP option with the trailing dot intact (e.g. `example.com.`). The DHCP client in systemd-networkd chops off the trailing dot during normalization. The hostname set by networkd therefore does not contain a trailing dot (and neither will it be persisted in the lease file).

I propose to follow the behavior of trimming a trailing dot/empty label when looking up a host's hostname via IMDS. This will pass current validation and be in line with what at least one other popular client does.

As an aside, given the above I think the trailing dot in the configured domain name of the VPC DHCP Option Set can be safely removed. This works around the problem in the current Bottlerocket 1.14.2 release.
